kubectl-cloudflow/cmd/deploy.go

```diff
@@ -81,32 +105,198 @@ the stored credentials.
+
+	for _, pulledImage := range pulledImages {
+		if pulledImage.Authenticated {
+			if err := verifyPasswordOptions(opts); err == nil {
+				if terminal.IsTerminal(int(os.Stdin.Fd())) && (opts.username == "" || opts.password == "") {
+					if !dockerConfigEntryExists(k8sClient, namespace, dockerRegistryURL) {
+						username, password := promptCredentials(dockerRegistryURL)
+						createOrUpdateImagePullSecret(k8sClient, namespace, dockerRegistryURL, username, password)
+					}
+				} else if opts.username != "" && opts.password != "" {
+					createOrUpdateImagePullSecret(k8sClient, namespace, dockerRegistryURL, opts.username, opts.password)
+				} else {
+					util.LogAndExit("Please provide username and password, by using both --username and --password-stdin, or, by using both --username and --password, or omit these flags to get prompted for username and password.")
+				}
+			} else {
+				util.LogAndExit("%s", err)
+			}
+		}
+	}
+
+	// Get the Cloudflow operator ownerReference
+	ownerReference := version.GetOwnerReferenceForCloudflowOperator()
+
+	serviceAccount := newCloudflowServiceAccountWithImagePullSecrets(namespace, ownerReference)
+
+	if _, err := createOrUpdateServiceAccount(k8sClient, namespace, serviceAccount); err != nil {
+		util.LogAndExit("%s", err)
+	}
+
+	// Delay the creation of the secret for after the ownerReferences has been added.
+	// Creating then updating the secret generates problem for the Flink streamlets deployment.
+	streamletNameSecretMap := deploy.CreateSecretsData(&applicationSpec, configurationKeyValues)
+	createOrUpdateStreamletSecrets(k8sClient, namespace, streamletNameSecretMap)
+
+	applicationSpec, err = copyReplicaConfigurationFromCurrentApplication(cloudflowApplicationClient, applicationSpec)
+	if err != nil {
+		util.LogAndExit("The application descriptor is invalid, %s", err.Error())
+	}
+
+	createOrUpdateCloudflowApplication(cloudflowApplicationClient, applicationSpec, ownerReference)
+
+	// When the CR has been created, create a ownerReference using the uid from the stored CR and
+	// then update secrets and service account with the ownerReference
+	storedCR, err := cloudflowApplicationClient.Get(applicationSpec.AppID)
+	if err != nil {
+		util.LogAndExit("Failed to retrieve the application `%s`, %s", applicationSpec.AppID, err.Error())
+	}
+	ownerReference = storedCR.GenerateOwnerReference()
+
+	streamletNameSecretMap = deploy.UpdateSecretsWithOwnerReference(ownerReference, streamletNameSecretMap)
+	createOrUpdateStreamletSecrets(k8sClient, namespace, streamletNameSecretMap)
+
+	serviceAccount.ObjectMeta.OwnerReferences = []metav1.OwnerReference{ownerReference}
+	if _, err := createOrUpdateServiceAccount(k8sClient, namespace, serviceAccount); err != nil {
+		util.LogAndExit("%s", err)
+	}
+
+	util.PrintSuccess("Deployment of application `%s` has started.\n", namespace)
+}
+
+func loadBlueprintConfigContent(contents string) (*configuration.Config, error) {
```

Comment:
Would be great if Blueprint had it's own struct that would get loaded into?

Reply:
Ah right but that will come with @skonto 's work... ok makes sense. (Sorry I'm just reading top down)

Reply:
yes, this will move out once we have @skonto 's code checked in. I tried to make my piece self containing - will do these refactorings before check-in. This will impose a serializable merge with Stavros' work, which I think, is logical.